### DISCUSSION TITLE
Not faster even slower but ...

### DISCUSSION THREAD

**Discussion by andre-goldmann** [modular/modular#890] [Ideas]
maybe this is just because I am new to mojo and still do not know enough about it. So here is my code:
Python
utils.py
`from datetime import timedelta
import pandas as pd

def printdelta(start, end):
    print(timedelta(seconds=end-start))

def loadData(file:str):
    df = pd.read_csv(file, delim_whitespace=True, header=0)  
    return df

def sliceDf(index, constant, df):
    return df[index:constant+index]`

calculation.py
`
def regression(lookback:int):
    start = timer()

    df = loadData('data/EURUSD_H1_201601040000_202309212100.csv')
    
    lsma_arr = []
    dates_arr = []

    for i in range(len(df) - 24):
        input_reg = df[i:25+i]
        x = pd.Series(range(len(input_reg.index))).values
        y = input_reg.CLOSE
        model = sm.OLS(y, sm.add_constant(x)).fit()
        pred = model.predict()[-1]
        lsma_arr.append(pred)
        dates_arr.append(input_reg.iloc[-1].name)

    lsma_df = pd.DataFrame({'LSMA': lsma_arr}, index=dates_arr)

    all_df = pd.concat([lsma_df,df], axis=1)
    all_df.dropna(inplace=True)
    
    end = timer()
    print(timedelta(seconds=end-start))

    return {"startTime": all_df.iloc[-lookback].DATE + " " + all_df.iloc[-lookback].TIME,
            "endTime": all_df.iloc[-1].DATE + " " + all_df.iloc[-1].TIME,
            "startValue": all_df.iloc[-lookback].LSMA,
            "endValue": all_df.iloc[-1].LSMA}`

calculation.mojo
`
fn regression(lookback:Int):
    try:
        Python.add_to_path(".")
        let timeit = Python.import_module("timeit")    

        let start = timeit.default_timer()

        let pd = Python.import_module("pandas")    
        let builtins = Python.import_module("builtins")
        let utils = Python.import_module("utils")
        let sm = Python.import_module("statsmodels.api")
        
        let df =  utils.loadData('data/EURUSD_H1_201601040000_202309212100.csv')   

        let lsma_arr: PythonObject = []
        let dates_arr: PythonObject = []

        for i in range(builtins.len(df) - 24):
            let input_reg = utils.sliceDf(i, 25, df) #df[i:25+i]
            let x = pd.Series(builtins.range(builtins.len(input_reg.index))).values
            let y = input_reg.CLOSE
            let model = sm.OLS(y, sm.add_constant(x)).fit()
            let pred = model.predict()[-1]
            lsma_arr.append(pred)
            dates_arr.append(input_reg.iloc[-1].name)
        
        #lsma_df = pd.DataFrame({'LSMA': lsma_arr}, index=dates_arr)
        #let lsma_df = utils.createDataFrame(lsma_arr, dates_arr)
        #all_df = pd.concat([lsma_df,df], axis=1)
        #all_df.dropna(inplace=True)

        #lsma_df = pd.DataFrame({'LSMA': lsma_arr}, index=dates_arr)                
        let end = timeit.default_timer()
        _ = utils.printdelta(start, end)
    except e:
        print("failed to show plot:", e.value)
        try:
            let traceback = Python.import_module("traceback")
            _ = traceback.print_exc()
        except e1:    
            print("Could not load traceback")`

Python-Took: 0:00:12.351395
Mojo-Took: 0:00:15.066987

So maybe someone finds the time and could explain/show how to improve this. Another question how I can the whole stacktrace, as I you can see I was trying to use python here. But this also shows not sufficient in case of an error.

**Comment by melonc2017** [comment]
maybe try catch

**Comment by gujuriu** [comment]
@andre-goldmann can you share a link to `EURUSD_H1_201601040000_202309212100.csv` file?

**Comment by gryznar** [comment]
I guess, that the slowdown is due to the fact that most job is done on Python side. Mojo uses CPython interpreter to execute Python code and that may cost.

**Comment by andre-goldmann** [comment]
@gujuriu the content of file looks like this:
DATE	TIME	OPEN	HIGH	LOW	CLOSE	TICKVOL	VOL	SPREAD
2016.01.04	00:00:00	1.08696	1.08723	1.08656	1.08665	904	629242000	7
2016.01.04	01:00:00	1.08664	1.08665	1.08449	1.08514	3093	3080554000	7
2016.01.04	02:00:00	1.08514	1.08575	1.08428	1.08451	3128	3278745000	7
2016.01.04	03:00:00	1.08450	1.08614	1.08270	1.08594	5611	5887393000	7
2016.01.04	04:00:00	1.08595	1.08755	1.08580	1.08713	5702	6175403000	7
2016.01.04	05:00:00	1.08712	1.08725	1.08667	1.08710	3042	3102477000	7
2016.01.04	06:00:00	1.08709	1.08833	1.08690	1.08791	3920	5448448000	7
2016.01.04	07:00:00	1.08790	1.09046	1.08737	1.08970	5150	6451089000	7
2016.01.04	08:00:00	1.08970	1.09021	1.08841	1.08887	4428	5337301000	7
2016.01.04	09:00:00	1.08887	1.09200	1.08814	1.09122	9731	11713469000	7
2016.01.04	10:00:00	1.09126	1.09462	1.09103	1.09239	12015	15491144000	7
2016.01.04	11:00:00	1.09243	1.09300	1.09124	1.09288	7748	10307525000	7

It is just an example, within the file there are around 48000 lines
